drivers/spi/spi_ll_stm32.c

```diff
@@ -340,6 +340,47 @@ static int spi_stm32_get_err(SPI_TypeDef *spi)
 }
 
 /* Shift a SPI frame as master. */
+#ifdef CONFIG_SPI_STM32_FIFO
+static void spi_stm32_shift_m(SPI_TypeDef *spi, struct spi_stm32_data *data)
+{
+	unsigned int frame_bytes = bits2bytes(
+		SPI_WORD_SIZE_GET(data->ctx.config->operation));
+	/*
+	 * Store data->MEMBERS in local variables (more likely to be stored in
+	 * CPU registers), mainly for performance benefits.
+	 */
+	size_t tx_fifo_room = data->tx_fifo_room / frame_bytes;
+	size_t transfer_len = data->transfer_len;
+	bool retry_tx = true;
+
+	while (retry_tx) {
+		retry_tx = false;
+
+		while (transfer_len > 0 &&
```

Comment:
Writing-then-reading frames in a loop here was discussed [here](https://github.com/zephyrproject-rtos/zephyr/pull/63173#discussion_r1395659932).

I had problems related to the fact that this driver supports different sizes for TX and RX buffers. Here in particular, I think if RX is bigger than TX, the last frames of RX won't be read as TX will have finished and so it won't send dummy zeroes to generate more clock cycles and so be able to read the last RX frames.

I also think I remember having problems related to being inside in the sending loop, then receive some frames to read, then seeing overrun errors because the "send loop" doesn't read. I added a `if (rx_fifo_not_empty()) { read(); break; }` to fix this at some point, but I don't remember the exact circumstances under which the problem appeared. 

Of course, the 2 problems above can be solved by adding more logic, but that adds complexity and I don't see a benefit that justifies so.

Also, why is `tx_fifo_room` required? Why not just write until the fifo is full?

In any case, have you run the SPI loopback tests? I wouldn't merge any change that hasn't passed (at least) those tests.

Could you provide your test plan, please?